core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala

```diff
@@ -141,7 +141,8 @@ class GroupMetadataManager(val brokerId: Int,
       // if we crash or leaders move) since the new leaders will still expire the consumers with heartbeat and
       // retry removing this group.
       val groupPartition = partitionFor(group.groupId)
-      val tombstone = new Message(bytes = null, key = GroupMetadataManager.groupMetadataKey(group.groupId))
+      val tombstone = new Message(bytes = null, key = GroupMetadataManager.groupMetadataKey(group.groupId),
+        timestamp = SystemTime.milliseconds, magicValue = Message.MagicValue_V1)
```

Comment:
If the message format is tied to the inter-broker protocol, the first broker that upgrades its inter-broker protocol to 0.10.0  will start to have v1 messages in its log. Since the rest of the brokers' inter-broker protocol is still before 0.10.0, they can only send v1 fetch request. Therefore, the first broker has to convert v1 messages down to v0 in the fetch response.

If we decouple the message format from inter-broker protocol, we can first keep message format to be v0. After every broker upgrades inter-broker protocol to 0.10.0 and starts using v2 fetch request, we then change message format to v1. At that point, there is no message format conversion needed.